### PR TITLE
MM-39351: fixes LastRootPostAt being null

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

After upgrading to v6.1 there is a chance for LastRootPostAt to be null,
this might be due to a faulty migration when upgrading to Version5350.
Version5350 had rootCountMigration where LastRootPostAt was used as a
temporary column which at the end was deleted.
Then on Version610 we are adding that column again, this time to keep.

If the migration to Version5350 had failed and it didn't drop the column
the Version610 migration would leave some channels with LastRootPostAt
being NULL.

This resulted in crashing the app.

This commit fixes that by running again a migration to again set
LastRootPostAt only when it's NULL this time and then set it to 0
for channels that might still have LastRootPostAt being NULL
(channels with no posts).

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-39351

#### Release Note

```release-note
Fixes scan error on column name "LastRootPostAt": converting NULL to int64
```

### CODE REVIEW

**Review by lieut-data — COMMENTED** [review]
Should we also harden the query itself to `COALESCE(LastRootPastAt, 0)`? This would ensure that bad data -- in the future, presumably -- wouldn't break channels at all, and we'd just fallback to safe semantics of starting over?

**Comment by koox00** [comment]
> Should we also harden the query itself to COALESCE(LastRootPastAt, 0)? This would ensure that bad data -- in the future, presumably -- wouldn't break channels at all, and we'd just fallback to safe semantics of starting over?

I was going to ask the same, if you think I need to change https://github.com/mattermost/mattermost-server/blob/009d02c57a923769dcb77acc32e57923f1f1d8f7/store/sqlstore/channel_store.go#L881 and add `COALESCE(LastRootPostAt, 0)`.

Well I'll go and change that then!

**Comment by streamer45** [comment]
Do we need that column to support NULL values? If the answer is no then adding a column default may be the cleanest approach.

**Comment by koox00** [comment]
> Do we need that column to support NULL values? If the answer is no then adding a column default may be the cleanest approach.

no we don't. The default should be 0. And I think we do add the column with a default value of 0

File: store/sqlstore/upgrade.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ import (
 const (
 	CurrentSchemaVersion   = Version610
```
Comment (streamer45) [inline store/sqlstore/upgrade.go:20]: We should update this to `Version611`.

**Review by streamer45 — COMMENTED** [review]
I am still unsure how this happened as I'd expect a sudden failure in a migration to prevent the schema version to get bumped hence the migration should be re-attempted next time around. But from what you are saying, the column was left there and the upgrade was allowed to proceed further.

**Comment by koox00** [comment]
> I am still unsure how this happened as I'd expect a sudden failure in a migration to prevent the schema version to get bumped hence the migration should be re-attempted next time around. But from what you are saying, the column was left there and the upgrade was allowed to proceed further.

I'm not sure what happened I'm just speculating.

**Comment by streamer45** [comment]
> I'm not sure what happened I'm just speculating.

I see. Personally I'd be a bit hesitant to fix something, especially inside a migration, until we have a clear picture of what could have caused it.

@lieut-data As far as you are aware, did this problem occur internally only or was it reported by customers as well?

As far as coalescing I am worried we may be hiding the real issue at play given the column should never host null values (it defaults to 0) unless something has gone really wrong.

**Comment by lieut-data** [comment]
> @lieut-data As far as you are aware, did this problem occur internally only or was it reported by customers as well?

@streamer45, as far as I understand there were two reports:
1) The Playbooks team ran into this in CI in a way that we solved by fixing our initialization data: https://community-daily.mattermost.com/core/pl/8ru81aqxepbcxdxa8fof3ezsbe.
2) A customer ran into this in production, it seems because the migration didn't initialize with the default value of `0` for channels that had no posts: https://community.mattermost.com/core/pl/6eyga4m6mibptfukmjzuekkigc.

I too am mystified why `NULL` values crept in with a `DEFAULT` of `0`.

**Comment by streamer45** [comment]
Thanks Jesse. Had a look at the original discussion (customer report) and if I understood correctly this is my best attempt to what could have cause this to happen:

1. 5.35 migration was somehow interrupted prior to completion which meant the temporary `Channels.LastRootPostAt` wasn't dropped. Moreover that column was originally created [without a default](https://github.com/mattermost/mattermost-server/blob/e464e4fb6222756bdcd061299540f4dc510f51d5/store/sqlstore/upgrade.go#L1168) so NULLs would have happened.
2. Server was restarted but the `hasMissingMigrationsVersion535` check passed as it [failed to check](https://github.com/mattermost/mattermost-server/blob/e464e4fb6222756bdcd061299540f4dc510f51d5/store/sqlstore/upgrade.go#L1279-L1294) for the column's presence. This resulted in the db schema version getting bumped and the following migrations to proceed.
3. Lastly, in 6.1 migration, the update query didn't take into account channels without posts due to the filtering on `Posts.RootId` which was misplaced as it should have been moved as part of the JOIN statement. This meant that some rows were left untouched (and with NULL in them).

@koox00 please correct me if I am wrong there.

/cc @isacikgoz for visibility

**Comment by agnivade** [comment]
IIUC, 5.35 adds a `LastRootAt`, not `LastRootPostAt`. So I am not sure how this can happen.

Looking at some more tables in our schema with bigint and a default value of 0, I can indeed see in my local DB that they don't have a default value. For example `sqlStore.CreateColumnIfNotExists("ThreadMemberships", "UnreadMentions", "bigint", "bigint", "0")`, the `ThreadMemberships.UnreadMentions` does not have a default value in my system.

And the channels.lastrootpostat has a default of '0'::bigint, which was strange.

Looking further, I discovered that the default column addition code always wraps the default value in single quotes. Which means something which should have been `alter table ThreadMemberships add column test bigint 0;` now becomes `alter table ThreadMemberships add column test bigint '0';`. This does not make a difference in MySQL, but in Postgres it does become `'0'::bigint`, instead of 0.

I am not sure if this makes any difference or not, but certainly it looks buggy and something that should be fixed naturally with the new migration system.

I am slightly inclined to let this happen after introducing our new migration system which should go live by this week. And then re-decide on how best do we want to proceed on this.

cc @isacikgoz as well.

**Comment by isacikgoz** [comment]
The theory of @streamer45 seems to be correct. I think the main problem is that the change of `LastRootPostAt` -> `LastRootAt` is not reflected into releases. Therefore it remained that way 🤦 
Until the last PR -> https://github.com/mattermost/mattermost-server/pull/19050/files which I realized it has been changed. I was not expecting a change in the previous migrations so probably it escaped from me while merging the branches.

**Comment by agnivade** [comment]
Thanks @isacikgoz for pointing out https://github.com/mattermost/mattermost-server/pull/18366. I was just looking at master, and did not realize that the migration itself was edited. Yes, in that case the above theory makes sense to me.

**Comment by koox00** [comment]
I can't remember what issue I was trying to solve when I edited the old migration, I realize know that it was a mistake.
I should have been more clear on the intention on that PR.

**Comment by amyblais** [comment]
Is this PR still needed for a dot release?

**Comment by koox00** [comment]
@amyblais I think the decision is that we are going to put in on hold until the new migrations are done

**Comment by agnivade** [comment]
The new migrations are in. Feel free to use the new model. @isacikgoz will be doing a demo soon with documentation update.

**Comment by koox00** [comment]
@agnivade sorry I got sidetracked and forgot about it, I'll start working on it

**Comment by agnivade** [comment]
Removing this from my queue for now. Feel free to request a review when it's ready.

**Comment by koox00** [comment]
closing this in favor of https://github.com/mattermost/mattermost-server/pull/19574

**Pull request closed without merging**